src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+    false,
+    3u,
+    1024u,
+    0u,
+    NULL,
+    NULL
+};
+
+int validate_context(dyad_ctx_t *ctx, bool check_cons, bool check_prod)
+{
+    if (!ctx || !ctx->h)
+    {
+        return DYAD_NOCTX;
+    }
+    if (check_cons && (ctx->cons_managed_path == NULL || strlen(ctx->cons_managed_path) == 0))
+    {
+        return DYAD_BADMANAGEDPATH;
+    }
+    if (check_prod && (ctx->prod_managed_path == NULL || strlen(ctx->prod_managed_path) == 0))
+    {
+        return DYAD_BADMANAGEDPATH;
+    }
+    return DYAD_OK;
+}
+
+int gen_path_key (const char* str, char* path_key, const size_t len,
+        const uint32_t depth, const uint32_t width)
+{
+    static const uint32_t seeds [10]
+        = {104677, 104681, 104683, 104693, 104701,
+            104707, 104711, 104717, 104723, 104729};
+
+    uint32_t seed = 57;
+    uint32_t hash[4] = {0u}; // Output for the hash
+    size_t cx = 0ul;
+    int n = 0;
+
+    if (path_key == NULL || len == 0ul) {
+        return -1;
+    }
+    path_key [0] = '\0';
+
+    for (uint32_t d = 0u; d < depth; d++)
+    {
+        seed += seeds [d % 10];
+        // TODO add assert that str is not NULL
+        MurmurHash3_x64_128 (str, strlen (str), seed, hash);
+        uint32_t bin = (hash[0] ^ hash[1] ^ hash[2] ^ hash[3]) % width;
+        n = snprintf (path_key+cx, len-cx, "%x.", bin);
+        cx += n;
+        if (cx >= len || n < 0) {
+            return -1;
+        }
+    }
+    n = snprintf (path_key+cx, len-cx, "%s", str);
+    if (cx + n >= len || n < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+int dyad_init(bool debug, bool check, bool shared_storage,
+        unsigned int key_depth, unsigned int key_bins,
+        const char *kvs_namespace, const char *prod_managed_path,
+        const char *cons_managed_path, bool intercept, dyad_ctx_t **ctx)
+{
+    printf("In dyad_init\n");
+    if (*ctx != NULL)
+    {
+        if ((*ctx)->initialized)
+        {
+            // TODO Indicate already initialized
+            DPRINTF ((*ctx), "DYAD context already initialized\n");
+        }
+        else
+        {
+            **ctx = dyad_ctx_default;
+        }
+        return DYAD_OK;
+    }
+    printf("Allocating ctx\n");
+    *ctx = (dyad_ctx_t*) malloc(sizeof(struct dyad_ctx));
+    if (*ctx == NULL)
+    {
+        fprintf (stderr, "Could not allocate DYAD context!\n");
+        return DYAD_NOCTX;
+    }
+    printf("Setting ctx values that don't need allocation\n");
+    **ctx = dyad_ctx_default;
+    (*ctx)->debug = debug;
+    (*ctx)->check = check;
+    (*ctx)->shared_storage = shared_storage;
+    (*ctx)->key_depth = key_depth;
+    (*ctx)->key_bins = key_bins;
+    (*ctx)->intercept = intercept;
+    printf("Setting KVS namespace\n");
+    if (kvs_namespace == NULL)
+    {
+        fprintf (stderr, "No KVS namespace provided!\n");
+        // TODO see if we want a different return val
+        return DYAD_NOCTX;
+    }
+    (*ctx)->kvs_namespace = (char*) malloc(strlen(kvs_namespace)+1);
```

Comment:
Since kvs_namespace is const array, you can store the length in a const variable and reuse whenever you need call `strlen()` again. `strlen()` will walk through the entire buffer until it reaches the null-terminating character.

Reply:
I've done this for `kvs_namespace`. Would we like to do this same for the managed paths?

Reply:
> I've done this for `kvs_namespace`. Would we like to do this same for the managed paths?

Yes

Reply:
Done